### PR TITLE
Refactoring RequestInit to use a Builder Pattern

### LINKED ISSUE #22427: Use builder pattern with RequestInit
If RequestInit::new accepts all of the mandatory arguments and then the builder pattern is used for customizable options, the resulting code might be easier to match against specification text like https://html.spec.whatwg.org/multipage/media.html#media-elements.

### PR DESCRIPTION
**Pull request by lucasfantacuci**
If RequestInit::new accepts all of the mandatory arguments and then the builder pattern is used for customizable options, the resulting code might be easier to match against specification text like 

---
- [x] `./mach build -d` does not report any errors
- [x] `./mach test-tidy` does not report any errors
- [x] These changes fix #22427
- [x] These changes do not require tests because it is a code refactoring.

---
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/22521)

### CODE REVIEW

**Comment by lucasfantacuci** [comment]
I implemented the builder pattern as a non-consuming builder. ref: https://doc.rust-lang.org/1.0.0/style/ownership/builders.html

File: components/net_traits/request.rs [hunk 1]
```diff
@@ -198,6 +199,101 @@ impl Default for RequestInit {
             url_list: vec![],
         }
     }
+
+    pub fn method(mut self, method: Method) -> RequestInit {
```
Comment (KiChjang) [inline components/net_traits/request.rs]: This looks like it is a consuming builder instead?

Comment (lucasfantacuci) [reply components/net_traits/request.rs]: Yeah, sorry, I switched the names. xD

Well, Actually I didn't implemented a builder pattern, since it hasn't a Builder class. I've created a constructor, called new, and setters methods, that return the object. Using this approach, I don't need this Builder class, it will have more readability, since it hasn't an extra class. I'm not sure about its performance.

What you think? Can I use this approach? or you really want a builder?

Comment (KiChjang) [reply components/net_traits/request.rs]: What you just described is the Rust builder pattern. Classes do not exist in Rust, and I've often understood that `RequestInit` is already a builder struct for `Request`. Perhaps renaming it `RequestBuilder` would make it more clear and apparent?

Comment (lucasfantacuci) [reply components/net_traits/request.rs]: Sure

Comment (lucasfantacuci) [reply components/net_traits/request.rs]: It confused me a lot, whose name, I was thinking that you were looking for a builder to the TYPE RequestInit, I didn't figure out, that the builder was the RequestInit.
I gonna make the changes as soon as possible.

And how about the builder, do you want a  non-consuming builder or a consuming builder?

Comment (KiChjang) [reply components/net_traits/request.rs]: I don't have any opinions either way, but I think the guide you referenced has a preference to non-consuming builders?

File: components/script/fetch.rs [hunk 2]
```diff
@@ -99,30 +99,29 @@ fn from_referrer_to_referrer_url(request: &NetTraitsRequest) -> Option<ServoUrl>
...
-            .immutable()
-            .clone(),
-        referrer_url: from_referrer_to_referrer_url(&request),
-        referrer_policy: request.referrer_policy,
-        pipeline_id: request.pipeline_id,
-        redirect_mode: request.redirect_mode,
-        cache_mode: request.cache_mode,
-        ..NetTraitsRequestInit::default()
-    }
+    NetTraitsRequestInit::new(request.url())
```
Comment (jdm) [inline components/script/fetch.rs]: I would actually like to revert the change to this function, and remove the use of `..NetTraitsRequesInit::default()`. That will make the compiler catch when we add members to NetTraitsRequestInit and need to be sure they are initialized correctly.

**Review by jdm — CHANGES_REQUESTED** [review]
Sorry about the delay! Thanks for making these changes; they have made clear some interesting holes in our current network requests that I plan to file issues about!

**Comment by jdm** [comment]
```
error: unused import: `std::default::Default`
  --> components/net_traits/request.rs:11:5
   |
11 | use std::default::Default;
   |     ^^^^^^^^^^^^^^^^^^^^^
   |
   = note: `-D unused-imports` implied by `-D warnings`
```

**Comment by lucasfantacuci** [comment]
My apologies, I was away. I gonna retake this task and finish it.

File: components/script/fetch.rs (hunk 2 above)
Comment (lucasfantacuci) [reply components/script/fetch.rs]: Sorry, I didn't get you.

Comment (jdm) [reply components/script/fetch.rs]: I would like to revert this particular change: this method should construct a new NetTraitsRequestInit object with an explicit structvliteral that initializes every field (instead of using `..NetTraitsRequestInit::default()`).

Comment (lucasfantacuci) [reply components/script/fetch.rs]: What do you think about a method overload to the method to the BuilderClass new passing all the arguments?

Comment (lucasfantacuci) [reply components/script/fetch.rs]: What do you think now?

Comment (lucasfantacuci) [reply components/script/fetch.rs]: Then we will be able to construct this object without setting the properties directly removing the "pub" from them.

e.g.:
```
RequestBuilder {
    url: url
    ...
}
```
It will throw an error.

Comment (lucasfantacuci) [reply components/script/fetch.rs]: @jdm

Comment (jdm) [reply components/script/fetch.rs]: The advantage of allowing some code like `request_init_from_request` to create the RequestInit value directly is that it's more easy to understand code like `use_url_credentials: false` instead of a `false` value that is one of 20 arguments. I think it's fine to keep the field values of RequestInit public and build it directly, rather than adding a method.

Comment (lucasfantacuci) [reply components/script/fetch.rs]: yeah, i've got you.

**Comment by lucasfantacuci** [comment]
I messed up with the branch. Finally fixed it haha. 
Well, How about now?

**Comment by jdm** [comment]
@bors-servo r+
Looks like a good improvement!

File: components/script/fetch.rs [hunk 3]
```diff
@@ -24,7 +24,7 @@ use crate::task_source::TaskSourceName;
 use ipc_channel::ipc;
 use ipc_channel::router::ROUTER;
 use js::jsapi::JSAutoCompartment;
-use net_traits::request::RequestInit as NetTraitsRequestInit;
+use net_traits::request::RequestBuilder as NetTraitsRequestInit;
```
Comment (KiChjang) [inline components/script/fetch.rs]: This probably doesn't need the renaming anymore.

File: components/script/fetch.rs [hunk 4]
```diff
@@ -105,12 +105,14 @@ fn request_init_from_request(request: NetTraitsRequest) -> NetTraitsRequestInit
...
         body: request.body.clone(),
+        service_workers_mode: ServiceWorkersMode::All,
         destination: request.destination,
         synchronous: request.synchronous,
         mode: request.mode.clone(),
+        cache_mode: request.cache_mode,
         use_cors_preflight: request.use_cors_preflight,
         credentials_mode: request.credentials_mode,
-        use_url_credentials: request.use_url_credentials,
+        use_url_credentials: false,
```
Comment (KiChjang) [inline components/script/fetch.rs]: Why is this changed to false?

Comment (lucasfantacuci) [reply components/script/fetch.rs]: lack of attention. Srry.

Comment (lucasfantacuci) [reply components/script/fetch.rs]: It may be a lack of attention. Srry.

Comment (lucasfantacuci) [reply components/script/fetch.rs]: The default value is false.

**Comment by KiChjang** [comment]
@bors-servo r=jdm,KiChjang

Thanks!
